src/coreclr/jit/utils.cpp

```diff
@@ -2585,3 +2585,368 @@ int64_t GetSigned64Magic(int64_t d, int* shift /*out*/)
 }
 #endif
 }
+
+namespace CheckedOps
+{
+bool IntAddOverflows(int32_t firstAddend, int32_t secondAddend, bool unsignedAdd)
+{
+    int32_t result = firstAddend + secondAddend;
+
+    if (unsignedAdd)
+    {
+        if ((int64_t)(uint32_t)result != ((int64_t)(uint32_t)firstAddend + (int64_t)(uint32_t)secondAddend))
+        {
+            return true;
+        }
+    }
+    else if ((int64_t)result != ((int64_t)firstAddend + (int64_t)secondAddend))
+    {
+        return true;
+    }
+
+    return false;
+}
+
+bool IntSubOverflows(int32_t minuend, int32_t subtrahend, bool unsignedSub)
+{
+    int32_t result = minuend - subtrahend;
+
+    if (unsignedSub)
+    {
+        if ((int64_t)(uint32_t)result != ((int64_t)(uint32_t)minuend - (int64_t)(uint32_t)subtrahend))
+        {
+            return true;
+        }
+    }
+    else if ((int64_t)(int32_t)result != ((int64_t)minuend - (int64_t)subtrahend))
+    {
+        return true;
+    }
+
+    return false;
+}
+
+bool IntMulOverflows(int32_t firstFactor, int32_t secondFactor, bool unsignedMul)
+{
+    int32_t result = firstFactor * secondFactor;
+
+    if (unsignedMul)
+    {
+        if ((int64_t)(uint32_t)result != ((int64_t)(uint32_t)firstFactor * (int64_t)(uint32_t)secondFactor))
+        {
+            return true;
+        }
+    }
+    else if ((int64_t)(int32_t)result != ((int64_t)firstFactor * (int64_t)secondFactor))
+    {
+        return true;
+    }
+
+    return false;
+}
+
+bool LongAddOverflows(int64_t firstAddend, int64_t secondAddend, bool unsignedAdd)
+{
+    int64_t result = firstAddend + secondAddend;
```

Comment:
> won't it work for both long and int cases?

Agree it should work for both `Int` & `Long` cases if we take the `Long` version. One thing I am not sure of is how would `INT64_MIN` be replaced given we don't have `std::numeric_limits`.

I actually wanted to go with the templated version at first, but there were a few reasons I didn't:
1\) I wanted to be able to just copy & paste code from `gtFoldExprConst`, as it has been used since (probably) Jit v1 and is thus known to be correct (and to make the review easier, since the `CheckedOps` change for integral types was only a refactoring one).
2\) I also wanted to enforce the invariant that these functions can only be called with `int32_t` or `int64_t`, no small/unsigned types allowed, without the ugly "assert in the general case, specialize for the known cases".

The latter lead me to `std::is_same`, which in turn I assumed I cannot use. Looks like that might not be the case (or I might be able to polyfill it, although my template skills are lacking at best), so I will investigate what can be done here.